### PR TITLE
Bypass fragile git ssh wrapper unless really needed

### LINKED ISSUE #38104: Git Module GIT_SSH Support
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

git

##### ANSIBLE VERSION

```
ansible 2.4.2.0
  config file = /home/jonni83/.ansible.cfg
  configured module search path = [u'/home/jonni83/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, May  3 2017, 07:55:04) [GCC 4.8.5 20150623 (Red Hat 4.8.5-14)]

```

##### CONFIGURATION

DEFAULT_HOST_LIST(/home/jonni83/.ansible.cfg) = [u'/home/jonni83/ansible/hosts']
DEFAULT_ROLES_PATH(/home/jonni83/.ansible.cfg) = [u'/etc/ansible/roles', u'/usr/share/ansible/roles']

##### OS / ENVIRONMENT

N/A
##### SUMMARY

My organization uses a custom compiled version of SSH.  Git supports defining the SSH program via the [GIT_SSH environment variable](https://git-scm.com/book/en/v2/Git-Internals-Environment-Variables#_miscellaneous).  That environment variable is commonly used to point to a script which contains general options.  The ansible module has a similar option [`ssh_opts`](http://docs.ansible.com/ansible/latest/git_module.html#options) which removes/overwrites the environment variable.   As the option's name suggests, it only allows the user to specify options.  However, the implementation overwrites `$GIT_SSH` and disables the user from defining the binary or script.

I created a local patch for git.py that implements a `ssh_bin` option which supports overwriting the name of the program instead of changing the current logic.  I'm sure there are several good ways of supporting this squashed feature.
##### STEPS TO REPRODUCE

Try specifying the program in ssh_opts
     or
Try setting the GIT_SSH environment variable

```yaml
- git:
    repo: ssh://git@github.com/mylogin/hello.git
    dest: /home/mylogin/hello
    ssh_opts: /path/to/my/ssh
```

##### EXPECTED RESULTS

Ansible uses the custom ssh program.
##### ACTUAL RESULTS

Ansible uses the SSH program on the PATH.

```
fatal:  [server.being.used]:  FAILED: =>  {
```

### LINKED ISSUE #64673: The git module should provide warning if the ssh_opts parameter is supplied but the version of git is 2.2.3 or earlier
##### SUMMARY

The `ssh_opts` parameter relies on setting environment variables for git, such as GIT_SSH. Git did not recognize this environment variable until version 2.3.0 (see https://git-scm.com/docs/git/2.3.0), and many contemporary OSes (RHEL 7, CentOS7) include older versions of Git (1.8.3.1).

It would be useful to provide a warning in this case, similar to the warning for the `depth` parameter. E.g. "Your git version is too old to support the ssh_opts argument."

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME

git.py
source control, git module

##### ADDITIONAL INFORMATION

It would save users time when trying to troubleshoot failed git connections via SSH.

```yaml
- name: clone a git repo
  git:
    repo: git@github.com:cherdt/hello_ansible.git
    dest: /opt/hello_ansible
    version: master
    ssh_opts: "{{-o IdentityFile=github_id_rsa}}"
```

### LINKED ISSUE #69846: Git module: don't encourage MITM attacks
##### SUMMARY

The Git module encourages use of `StrictHostKeyChecking=no` with SSH, which makes the connection vulnerable to a MITM attack.

##### ISSUE TYPE

OpenSSH 7.5+ (2017-03-20) added a safer option which automatically accepts the key on first use but will not allow a MITM on subsequent connections: `StrictHostKeyChecking=accept-new` 

https://man.openbsd.org/ssh_config#StrictHostKeyChecking

It would be great if Ansible encouraged use of that mode by default in the documentation and to implement features such as the Git module's [`accept_hostkey`](https://github.com/ansible/ansible/blob/323d2adfcce9a53e906c2c4a40f6ac5fb32a0e68/lib/ansible/modules/git.py#L1103-L1108) feature.

##### ANSIBLE VERSION

```paste below
N/A
```

##### COMPONENT NAME
* `/modules/source_control/git.py`

### PR DESCRIPTION
**Pull request by bcoca**
git module now uses env vars exclusively

  - updated docs to clarify usage
  - now env vars append instead of overwrite to allow existing custom setups to keep working
    fixes #38104
    fixes #64673
    fixes #64674
  - added note for hostkeychecking more securely
    fixes #69846

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

modules/git.py

### CODE REVIEW

File: lib/ansible/modules/git.py [hunk 1]
```diff
@@ -43,23 +43,28 @@
...
         description:
-            - Creates a wrapper script and exports the path as GIT_SSH
-              which git then automatically uses to override ssh arguments.
-              An example value could be "-o StrictHostKeyChecking=no"
-              (although this particular option is better set by
-              I(accept_hostkey)).
+            - Options git will pass to ssh when used as protocol, it works via GIT_SSH_OPTIONS.
+            - For older versions it appends GIT_SSH_OPTIONS to GIT_SSH/GIT_SSH_COMMAND.
+            - Other options can add to this list, like I(key_file) and I(accept_hostkey).
+            - An example value could be "-o StrictHostKeyChecking=no" (although this particular
```
Comment (bmillemathias) [inline lib/ansible/modules/git.py:67]: The example could use a different option than one already covered above as you wrote , like `Compression=yes`. This part could also clarify if the module honors the content of ssh config file.

File: lib/ansible/modules/git.py [hunk 2]
```diff
@@ -43,23 +43,28 @@
         default: "HEAD"
     accept_hostkey:
         description:
-            - If C(yes), ensure that "-o StrictHostKeyChecking=no" is
-              present as an ssh option.
+            - Will ensure or not that "-o StrictHostKeyChecking=no" is present as an ssh option.
+            - Be aware that this disables a protection against MITM attacks.
+            - Those using OpenSSH >= 7.5 might want to set I(ssh_opt) to 'StrictHostKeyChecking=accept-new'
+              instead, it does not remove the MITM issue but it does restrict it to the first attempt.
```
Comment (bmillemathias) [inline lib/ansible/modules/git.py:49]: The other sections are not using bullet-points, for consistency could the update be written in the same manner ?

**Review by bmillemathias — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/git.py (hunk 2 above)
Comment (bcoca) [reply lib/ansible/modules/git.py:49]: not bullet points, description is always a list, this is data type. Also making very separate points.

File: lib/ansible/modules/git.py (hunk 1 above)
Comment (bcoca) [reply lib/ansible/modules/git.py:67]: kept original as it stresses how other options can also modify the opts.

**Pull request merged**